### ISSUE TITLE
Undocumented behaviour: %x uses .String() if it exists

### ISSUE THREAD

**Issue by someone235** [golang/go#46076]
The below program will print the hex encoded value of `a.String()`, and the hex encoded value of `b`. AFAIK, this is not documented anywhere.
```
package main

import (
	"fmt"
)

type a [1]byte
func (_ a) String() string {
	return "bla"
}

type b [1]byte

func main() {
	x := a{}
	y := b{}
	fmt.Printf("%x %x\n", x, y)
}

```

**Comment by davecheney** [comment]
I believe that this is documented, https://golang.org/src/fmt/doc.go#L49

**Comment by robpike** [comment]
Yes, it is documented. Also see https://blog.golang.org/strings

**Comment by someone235** [comment]
@robpike I don't find it. Can you please quote the relevant part?

**Comment by davecheney** [comment]
> A shorter way to generate presentable output for a messy string is to use the %x (hexadecimal) format verb of fmt.Printf. It just dumps out the sequential bytes of the string as hexadecimal digits, two per byte.

**Comment by seankhliao** [comment]
Stringer documents this more clearly
> The String method is used to print values passed as an operand to any format that accepts a string

**Comment by ianlancetaylor** [comment]
Quoting https://golang.org/pkg/fmt:

> String and slice of bytes (treated equivalently with these verbs):
> 
> %s	the uninterpreted bytes of the string or slice
> %q	a double-quoted string safely escaped with Go syntax
> %x	base 16, lower-case, two characters per byte
> %X	base 16, upper-case, two characters per byte

then later:

> If the format (which is implicitly %v for Println etc.) is valid for a string (%s %q %v %x %X), the following two rules apply:
> 
> 4. If an operand implements the error interface, the Error method will be invoked to convert the object to a string, which will then be formatted as required by the verb (if any).
> 
> 5. If an operand implements method String() string, that method will be invoked to convert the object to a string, which will then be formatted as required by the verb (if any).

**Comment by lpar** [comment]
Yes, but I just hit this behavior:

```go
package main

import (
	"fmt"
)

type MyType int64

func (m MyType) String() string {
	return "Really Go?"
}

const MyConstant MyType = 524287

func main() {
	fmt.Printf("%d = %x", MyConstant, MyConstant)
}
```

Having a constant value with an underlying type of `int64` print as a hex encoded version of its string representation, rather than the hex encoding of its value, was really unexpected to me, even though the type happened to be a Stringer. It wasn't until I happened to decode what looked like hex garbage into ASCII that I worked out what was going on.

I can understand why it makes sense for a byte array to get turned into a string first by `%x`, but an `int64`, not so much.
